presto-spi/src/main/java/com/facebook/presto/spi/relation/LambdaDefinitionExpression.java

```diff
@@ -111,4 +115,55 @@ private static void checkArgument(boolean condition, String message, Object... m
             throw new IllegalArgumentException(format(message, messageArgs));
         }
     }
+    private static class CanonicalizeExpression
+            implements RowExpressionVisitor<String, Void>
+    {
+        private final Map<String, String> canonicalizedArguments = new HashMap<>();
+
+        public CanonicalizeExpression(List<String> arguments, List<Type> argumentTypes)
+        {
+            for (int i = 0; i < arguments.size(); i++) {
+                canonicalizedArguments.put(arguments.get(i), format("%s_%d", argumentTypes.get(i).toString(), i));
+            }
+        }
+
+        @Override
+        public String visitCall(CallExpression call, Void context)
+        {
+            return format("%s(%s)", call.getDisplayName(), String.join(", ", call.getArguments().stream().map(e -> e.accept(this, null)).collect(Collectors.toList())));
+        }
+
+        @Override
+        public String visitInputReference(InputReferenceExpression reference, Void context)
+        {
+            return reference.toString();
+        }
+
+        @Override
+        public String visitConstant(ConstantExpression literal, Void context)
+        {
+            return format("%d", literal.hashCode());
+        }
+
+        @Override
+        public String visitLambda(LambdaDefinitionExpression lambda, Void context)
+        {
+            return format("(%s) -> %s", String.join(", ", lambda.argumentTypes.stream().map(Type::toString).collect(Collectors.toList())), lambda.body.accept(this, null));
+        }
+
+        @Override
+        public String visitVariableReference(VariableReferenceExpression reference, Void context)
+        {
+            if (canonicalizedArguments.containsKey(reference.getName())) {
```

Comment:
Will this handle duplicated variables with the same name in different scopes?
```java
SELECT
    TRANSFORM(elements, v -> REDUCE(v, 0, (s, v) -> s + v, s -> s)) AS multiplied_elements
FROM (
    VALUES
        (ARRAY[ARRAY[1, 2]]),
        (ARRAY[ARRAY[1, 2]]),
        (ARRAY[ARRAY[1, 2]])
) AS t(elements)

Reply:
we always assign a new variable in planning so the two `v`s will not have the same name in plan.

Reply:
@rongrong : One random thought, after canonicalization, argument name can be the same in different scopes right? For example the second BIGINT argument in different scope will always be canonicalized into `bigint_2`. Will this be handled correctly?

Reply:
> @rongrong : One random thought, after canonicalization, argument name can be the same in different scopes right? For example the second BIGINT argument in different scope will always be canonicalized into `bigint_2`. Will this be handled correctly?

Yes, the goal is to make two argument of the same type in different lambda the same. They are logically the same. You can invoke the same function (even if you generate two functions they will be the same anyways). The whole purpose of this is to enable common sub expressions for lambdas.